lib/synapse/service_watcher/marathon.rb

```diff
@@ -0,0 +1,105 @@
+require "synapse/service_watcher/base"
+require 'json'
+require 'resolv'
+
+module Synapse
+  class MarathonWatcher < BaseWatcher
+    def start
+      @check_interval = @discovery['check_interval'] || 10.0
+      @connection = nil
+      @watcher = Thread.new { watch }
+    end
+
+    def stop
+      @connection.finish
+    rescue
+      # pass
+    end
+
+  private
+
+    def validate_discovery_opts
+      required_opts = %w[marathon_api_url application_name]
+
+      required_opts.each do |opt|
+        if @discovery.fetch(opt, '').empty?
+          raise ArgumentError,
+            "a value for services.#{@name}.discovery.#{opt} must be specified"
+        end
+      end
+    end
+
+    def attempt_marathon_connection
+      @marathon_api = URI.join(@discovery['marathon_api_url'], "/v2/apps/#{@discovery['application_name']}/tasks")
+
+      begin
+        @connection = Net::HTTP.new(@marathon_api.host, @marathon_api.port)
+        @connection.open_timeout = 5
+        @connection.start
+      rescue => ex
+        @connection = nil
+        log.error "synapse: could not connect to marathon at #{@marathon_api}: #{ex}"
+
+        raise ex
+      end
+    end
+
+    def watch
+      until @should_exit
+        retry_count = 0
+        start = Time.now
+
+        begin
+          if @connection.nil?
+            attempt_marathon_connection
+          end
+
+          req = Net::HTTP::Get.new(@marathon_api.request_uri)
+          req['Accept'] = 'application/json'
+          response = @connection.request(req)
+
+          tasks = JSON.parse(response.body).fetch('tasks', [])
+          backends = tasks.keep_if { |task| task['startedAt'] }.map do |task|
+            {
+              'name' => task['host'],
+              'host' => task['host'],
+              'port' => task['ports'].first,
+            }
+          end.sort_by { |task| task['name'] }
+
+          if backends.empty?
+            log.warn "synapse: no backends discovered for #{@discovery['application_name']}"
+          else
+            previous_backends = @backends
+            set_backends(backends)
+            new_backends = @backends
+
+            unless previous_backends == new_backends
+              log.info "synapse: found #{backends.length} backends for #{@discovery['application_name']}"
+
+              reconfigure!
+            end
+          end
+        rescue EOFError
```

Comment:
Do we need to set @connection to nil here to ensure we attempt to connect back to marathon?